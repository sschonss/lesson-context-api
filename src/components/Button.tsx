import React from "react";
import { Button as RnButton, StyleSheet, View } from "react-native";

export default function Button({
  title,
  onPress,
  color,
}: {
  title: string;
  onPress: () => void;
  color: string;
}) {
  return (
    <View style={styles.container}>
      <RnButton title={title} onPress={onPress} color={color} />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
  },
});
