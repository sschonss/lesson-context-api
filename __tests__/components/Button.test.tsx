import { describe, expect, test } from "@jest/globals";
import React from "react";
import renderer from "react-test-renderer";
import Button from "../../src/components/Button";
import { fireEvent } from "@testing-library/react-native";

describe("Button", () => {
  test("should title text be 'Prev'", () => {
    const button = renderer.create(
      <Button title="Prev" onPress={() => {}} color="red" />
    );

    const instance = button.root;
    const buttonElement = instance.findByType(Button);
    expect(buttonElement.props.title).toBe("Prev");
  });

  test("should color be 'red'", () => {
    const button = renderer.create(
      <Button title="Prev" onPress={() => {}} color="red" />
    );

    const instance = button.root;
    const buttonElement = instance.findByType(Button);
    expect(buttonElement.props.color).toBe("red");
  });

  test("should call onPress", () => {
    const onPress = jest.fn();
    const button = renderer.create(
      <Button title="Prev" onPress={onPress} color="red" />
    );

    const instance = button.root;
    const buttonElement = instance.findByType(Button);
    buttonElement.props.onPress();
    expect(onPress).toBeCalled();
  });

  test("should match snapshot", () => {
    const button = renderer.create(
      <Button title="Prev" onPress={() => {}} color="red" />
    );

    expect(button.toJSON()).toMatchSnapshot();
  });

  test("should match snapshot with different props", () => {
    const button = renderer.create(
      <Button title="Next" onPress={() => {}} color="blue" />
    );

    expect(button.toJSON()).toMatchSnapshot();
  });

  test("mock function should be called", () => {
    const onPress = jest.fn(() => console.log("Button pressed"));
    const button = renderer.create(
      <Button title="Next" onPress={onPress} color="blue" />
    );

    const instance = button.root;
    const buttonElement = instance.findByType(Button);
    buttonElement.props.onPress();
    expect(onPress).toBeCalled();
  });

  test("pressing button should call onPress", () => {
    const onPress = jest.fn(() => console.log("Button pressed"));
    const button = renderer.create(
      <Button title="Next" onPress={onPress} color="blue" />
    );

    const instance = button.root;
    const buttonElement = instance.findByType(Button);
    fireEvent.press(buttonElement);
    expect(onPress).toBeCalled();
  });

  test("should throw error", () => {
    const onPress = jest.fn(() => {
      throw new Error("Error");
    });

    const button = renderer.create(
      <Button title="Next" onPress={onPress} color="blue" />
    );

    const instance = button.root;
    const buttonElement = instance.findByProps({ title: "Next" });

    expect(() => buttonElement.props.onPress()).toThrowError();
  });
});
