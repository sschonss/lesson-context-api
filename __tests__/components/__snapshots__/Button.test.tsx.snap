// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button should match snapshot 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            {
              "color": "red",
            },
          ]
        }
      >
        Prev
      </Text>
    </View>
  </View>
</View>
`;

exports[`Button should match snapshot with different props 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
      "justifyContent": "center",
    }
  }
>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
            {
              "color": "blue",
            },
          ]
        }
      >
        Next
      </Text>
    </View>
  </View>
</View>
`;
